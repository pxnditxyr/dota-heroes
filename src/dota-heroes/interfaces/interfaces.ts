export interface Hero {
    id: string;
    name: string;
    another_name: string;
    attribute: string;
    attack: string;
    information: string;
};
