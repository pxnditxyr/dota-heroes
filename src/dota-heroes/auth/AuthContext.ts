import { createContext } from 'react';

import { User } from '../interfaces/interfaces';

export type AuthContextProps = {
    user: User;
    authLogin: ( user : User ) => void;
    authLogout: ( logged : boolean ) => void;
};

export const AuthContext = createContext<AuthContextProps>( {} as AuthContextProps );
