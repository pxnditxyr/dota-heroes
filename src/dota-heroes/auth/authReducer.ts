import { User } from '../interfaces/interfaces';

type authAction =
    | { type: "authLogin", payload: User }
    | { type: "authLogout", payload: { logged : boolean } }

export const authReducer = ( state : User, action : authAction ) : User => {

    switch ( action.type ) {
        case "authLogin":
            return {
                ...action.payload,
                logged: true,
            }

        case "authLogout":
            return {
                ...state,
                logged: false
            }
        default: 
            return state;
    }
};
