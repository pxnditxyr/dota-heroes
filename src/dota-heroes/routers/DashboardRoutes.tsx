import {
    Redirect,
    Route,
    Switch,
} from 'react-router-dom';

import { Navbar } from '../components/ui/Navbar';
import { StrengthScreen } from '../components/strength/StrengthScreen';
import { AgilityScreen } from '../components/agility/AgilityScreen';
import { IntelligenceScreen } from '../components/intelligence/IntelligenceScreen';
import { SearchScreen } from '../components/search/SearchScreen';
import { HeroScreen } from '../components/heroes/HeroScreen';

export const DashboardRoutes = () => {
    return (
        <>
            <Navbar />
            <Switch>
                <Route exact path="/strenght" component={ StrengthScreen } />
                <Route exact path="/agility" component={ AgilityScreen } />
                <Route exact path="/intelligence" component={ IntelligenceScreen } />
                <Route exact path="/search" component={ SearchScreen } />
                <Route exact path="/hero/:heroId" component={ HeroScreen } />
                <Redirect to="/strenght" />
            </Switch>
        </>
    )
}
