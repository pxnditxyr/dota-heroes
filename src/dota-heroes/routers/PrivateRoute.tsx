import { Navigate, useLocation } from 'react-router-dom';
import { useAuth } from '../hooks/useAuth';

interface Props {
    children: JSX.Element | JSX.Element[];
}

export const PrivateRoute = ( { children } : Props  ) => {

    const { user } = useAuth();

    const { pathname, search } = useLocation();

    localStorage.setItem( 'lastPath', pathname + search );

    return user.logged
        ? <> { children } </>
        : <Navigate to="login" />
};
