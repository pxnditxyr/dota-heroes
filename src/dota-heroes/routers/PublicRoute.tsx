import { Navigate } from 'react-router';
import { useAuth } from '../hooks/useAuth';

interface Props {
    children : JSX.Element | JSX.Element[];
}

export const PublicRoute = ( { children } : Props ) => {

    const { user } = useAuth();

    return user.logged
        ? <Navigate to="/" />
        : <> { children } </>
}
