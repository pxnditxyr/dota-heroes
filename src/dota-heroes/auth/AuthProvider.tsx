import { useEffect, useReducer } from 'react';
import { User } from '../interfaces/interfaces';

import { AuthContext } from './AuthContext';
import { authReducer } from './authReducer';


const INITIAL_STATE : User = {
    id: "",
    name: "",
    email: "",
    password: "",
    logged: false,
};

const init : () => User = () => {
    return JSON.parse( localStorage.getItem( 'user' ) || "{}" );
}

interface AuthProviderProps {
    children: JSX.Element | JSX.Element[];
};

export const AuthProvider = ( { children }: AuthProviderProps ) => {

    const [ user, dispatch ] = useReducer( authReducer, INITIAL_STATE, init );
    useEffect( () => {
        if ( !user ) return;
        localStorage.setItem( 'user', JSON.stringify( user ) );
    }, [ user ] )

    const authLogin = ( user: User ) => {
        dispatch({ type: "authLogin", payload: user });
    };

    const authLogout = ( logged : boolean ) => {
        dispatch({ type: "authLogout", payload: { logged } });
    };
    const auth = () => {
        dispatch({ type: "auth", payload: {} })
    }

    return (
        <AuthContext.Provider
            value={{
                user,
                authLogin,
                authLogout,
                auth,
            }}
        >
            { children }
        </AuthContext.Provider>
    );
};
