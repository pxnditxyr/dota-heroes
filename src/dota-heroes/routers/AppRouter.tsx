import {
    BrowserRouter as Router,
    Route,
    Switch
} from 'react-router-dom';

import { DashboardRoutes } from './DashboardRoutes';
import { LoginScreen } from '../components/login/LoginScreen';

export const AppRouter = () => {
    return (
        <Router>
            <Switch>
                <Route exact path="/login" component={ LoginScreen } />
                <Route path="/" component={ DashboardRoutes } />
            </Switch>
        </Router>
    );
};
