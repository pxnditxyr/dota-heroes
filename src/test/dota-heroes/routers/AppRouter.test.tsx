import { mount } from 'enzyme';
import { AppRouter } from '../../../dota-heroes/routers/AppRouter';
import { AuthContext } from '../../../dota-heroes/auth/AuthContext';
import { User } from '../../../dota-heroes/interfaces/interfaces';

describe( "Tests in <AppRouter />", () => {


    test( "If is not authenticated, must show the login", () => {
        const contextValue = {
            user: {
                id: "",
                name: "",
                email: "",
                password: "",
                logged: false,
            },
            authLogin: ( user : User ) => { console.log( user ) },
            authLogout: ( logged : boolean ) => { console.log( logged ) },
            auth: () => { console.log( "a" ) },
        };
        const wrapper = mount( 
            <AuthContext.Provider value={ contextValue }>
                <AppRouter />
            </AuthContext.Provider>
        );
        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find( 'h1' ).text().trim() ).toBe( "Login" );
    });

    test( "If is authenticated, must show the StrenghtScreen", () => {
        const contextValue = {
            user: {
                id: "1",
                name: "testName",
                email: "example@gmail.com",
                password: "MD5_testPassword",
                logged: true,
            },
            authLogin: ( user : User ) => { console.log( user ) },
            authLogout: ( logged : boolean ) => { console.log( logged ) },
            auth: () => { console.log( "a" ) },
        };
        
        const wrapper = mount(
            <AuthContext.Provider value={ contextValue }>
                <AppRouter />
            </AuthContext.Provider>
        );
        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find( 'h1' ).text().trim() ).toBe( "Strength" );

    });
});
