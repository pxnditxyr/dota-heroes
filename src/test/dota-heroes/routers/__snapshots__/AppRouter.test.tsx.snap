// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <AppRouter /> If is authenticated, must show the StrenghtScreen 1`] = `
Array [
  " ",
  Array [
    <nav>
      <div>
        <a
          className=""
          href="/strenght"
          onClick={[Function]}
        >
          Strenght
        </a>
        <a
          className=""
          href="/agility"
          onClick={[Function]}
        >
          Agility
        </a>
        <a
          className=""
          href="/intelligence"
          onClick={[Function]}
        >
          Intelligence
        </a>
      </div>
      <div>
        <a
          className=""
          href="/search"
          onClick={[Function]}
        >
          Search
        </a>
        <span>
           
          testName
           
        </span>
        <button
          onClick={[Function]}
        >
          Logout
        </button>
      </div>
    </nav>,
    Array [
      <h1>
         Strength 
      </h1>,
      <hr />,
      <ul>
        <article
          className="card small"
        >
          <section
            className="card__header"
          >
            <img
              alt="strength-abaddon"
              src="./assets/images/heroes/strength-abaddon.jpg"
            />
          </section>
          <section
            className="card__body"
          >
            <h3>
               
              Abaddon
               
            </h3>
            <span>
               
              Lord of Avernus
               
            </span>
            <p>
               
              Strength
               
            </p>
            <p>
               
              meele
               
            </p>
          </section>
          <section
            className="card__footer"
          >
            <p>
               
              Abaddon, the Lord of Avernus, is a melee strength name known as one of the most versatile characters in Dota due to his rather low mana dependence, short spell cooldowns and a large number of viable item choices. His ability to help sustain his allies and himself plus his strong tower diving capacity give him solid lane presence. Many of his abilities offer a large sum of utility, which makes him a strong support name. Mist Coil serves as both a single target nuke and heal that helps shift the sustainability of both allied and enemy namees in a lane at his will, though sacrificing a portion of his own health. Aphotic Shield holds as one of the most useful abilities in the game, able to shield a target from some damage while also able to reflect said damage to a huge area. The most important aspect is how it is able to dispel many status effects such as slows and stuns. His other abilities allow him to become a mix between a semi-carry and tank. Abaddon's passive, Curse of Avernus, allows his attacks to not only slow down his enemy, but also increase the attack speed of any ally attacking the same target while also silencing the target after a set of attacks. With his ultimate, Borrowed Time, Abaddon is able to shift all non-HP-removal damage he receives into health. When not on cooldown, Borrowed Time may activate passively when his health falls under a certain threshold, even under the most dire situations. Due to his powerful spells and versatility, Abaddon is an excellent addition to any team.
               
            </p>
            <a
              href="/hero/strength-abaddon"
              onClick={[Function]}
            >
              More
            </a>
          </section>
        </article>
        <article
          className="card small"
        >
          <section
            className="card__header"
          >
            <img
              alt="strength-alchemist"
              src="./assets/images/heroes/strength-alchemist.jpg"
            />
          </section>
          <section
            className="card__body"
          >
            <h3>
               
              Alchemist
               
            </h3>
            <span>
               
              Razzil Darkbrew
               
            </span>
            <p>
               
              Strength
               
            </p>
            <p>
               
              meele
               
            </p>
          </section>
          <section
            className="card__footer"
          >
            <p>
               
              Razzil Darkbrew, the Alchemist, is a melee strength name whose alchemical prowess makes him a strange but versatile fighter. He is an unusual carry based upon transmuting killed creeps into large amounts of bonus gold, with both an early game and late game presence due to the enormous health regeneration from his ultimate and the first strike nature of his other spells. His low but balanced attributes and the sure promise of gold for items means he can be one of the most disparately built namees in the game. Unstable Concoction is his main contribution early on, dealing good damage and a lengthy stun. Acid Spray allows him to rapidly clear waves of creeps for his Greevil's Greed to contribute massive amounts of extra income. A well-equipped Alchemist can then use Chemical Rage to its fullest effect, as the incredible regeneration and base attack time reduction make it one of the best steroid abilities in the game.
               
            </p>
            <a
              href="/hero/strength-alchemist"
              onClick={[Function]}
            >
              More
            </a>
          </section>
        </article>
        <article
          className="card small"
        >
          <section
            className="card__header"
          >
            <img
              alt="strength-axe"
              src="./assets/images/heroes/strength-axe.jpg"
            />
          </section>
          <section
            className="card__body"
          >
            <h3>
               
              Axe
               
            </h3>
            <span>
               
              Mogul Khan
               
            </span>
            <p>
               
              Strength
               
            </p>
            <p>
               
              meele
               
            </p>
          </section>
          <section
            className="card__footer"
          >
            <p>
               
              Mogul Khan, the Axe, is a melee strength name, infamous for creating chaos in battle and thriving off of it. Axe is commonly played as an initiator, as his potent disables can disrupt the enemy's formation and give his team openings in fights. Axe's fighting style demands that he gets up close and very personal, as he can taunt enemies into attacking him with Berserker's Call, preventing them from engaging his teammates and granting him a huge amount of bonus armor. Once he has the enemy's attention, he can counter those who strike him with a sweeping Counter Helix that slashes all enemies at melee range with pure damage. He also has a slow and damage-over-time nuke in Battle Hunger, which can harass the enemy or force them to join in battle, while also slowing their movement speed and boosting his own. Finally, he has the ability to unconditionally remove weakened enemies from the battlefield if they are under a certain threshold with Culling Blade, while simultaneously boosting his allies' speed. A bloodthirsty and indomitable warrior, Axe's presence on the battlefield is felt by friend and foe alike, dominating the attention of the enemy wherever he goes.
               
            </p>
            <a
              href="/hero/strength-axe"
              onClick={[Function]}
            >
              More
            </a>
          </section>
        </article>
      </ul>,
    ],
  ],
  " ",
]
`;

exports[`Tests in <AppRouter /> If is not authenticated, must show the login 1`] = `
Array [
  " ",
  Array [
    <h1>
       Login 
    </h1>,
    <hr />,
    <button
      onClick={[Function]}
    >
      Login
    </button>,
  ],
  " ",
]
`;
